src/backends/github/implementation.js

```diff
@@ -29,11 +30,20 @@ export default class GitHub {
 
   authenticate(state) {
     this.token = state.token;
-    this.api = new API({ token: this.token, branch: this.branch, repo: this.repo });
-    return this.api.user().then((user) => {
-      user.token = state.token;
-      return user;
-    });
+    this.api = new API({ token: this.token, branch: this.branch, repo: this.repo, api_root: this.api_root });
+    return this.api.user().then(user =>
+      this.api.collaborator(user.login).then((status) => {
+        if (status === 404 || status === 403) {
+          // Unauthorized user
+          throw new Error("Your GitHub user account does not have access to this repo.");
+        } else if (status === 204) {
+          // Authorized user
+          user.token = state.token;
+          return user;
+        }
```

Comment:
The preceding status-checking logic should move to `collaborator`/`isCollaborator` (whatever that ends up being named) instead of being handled here - the only part of this that's relevant to authentication is whether the user is a collaborator or not, not the specific status code. This is part of the more detailed refactor I suggested above.